src/threadpool.c

```diff
@@ -57,104 +59,132 @@ static void uv__cancelled(struct uv__work* w) {
 }
 
 
+/* Only call when you hold NO locks. */
+static void do_work(struct uv__work* w) {
+  w->work(w);
+  uv_mutex_lock(&w->loop->wq_mutex);
+  w->work = NULL;  /* Signal uv_cancel() that the work req is done executing. */
+  QUEUE_INSERT_TAIL(&w->loop->wq, &w->wq);
+  uv_async_send(&w->loop->wq_async);
+  uv_mutex_unlock(&w->loop->wq_mutex);
```

Comment:
It should be possible to avoid taking out the lock here when running in synchronous (i.e. zero threads) mode.